// coc-jsref/src/fixers/experimental-jest-revert-only.js
"use strict";
exports.__esModule = true;
var Range = require("../range");
var Ast = require("../ast");
var Patch = require("../patch");
var fixer = {
    suggestCodeAction: function (params) {
        var j = params.j, ast = params.ast;
        var node = Ast.findFirstNode(ast, j.CallExpression, function (n) {
            return (n.loc !== null &&
                Range.isInside(params.selection, n.loc) &&
                j.MemberExpression.check(n.callee) &&
                j.Identifier.check(n.callee.object) &&
                (n.callee.object.name === 'it' || n.callee.object.name === 'test') &&
                j.Identifier.check(n.callee.property) &&
                n.callee.property.name === 'only');
        });
        if (!node ||
            !node.loc ||
            !j.MemberExpression.check(node.callee) ||
            !j.Identifier.check(node.callee.object)) {
            return null;
        }
        var functionName = node.callee.object.name;
        return {
            title: "Remove \"only\"",
            data: {
                position: node.loc.start,
                functionName: functionName
            }
        };
    },
    createEdit: function (params) {
        var data = params.data, ast = params.ast, j = params.j;
        var node = Ast.findFirstNode(ast, j.CallExpression, function (n) { return Ast.isOnPosition(n, data.position); });
        if (!node) {
            return null;
        }
        var id = node.callee;
        var newNode = j.identifier(data.functionName);
        return Patch.replaceNode(j, id, newNode);
    }
};
exports["default"] = fixer;
