// coc-jsref/src/range.js
"use strict";
exports.__esModule = true;
exports.format = exports.isInsideStrict = exports.isInside = exports.toProtocol = exports.fromProtocolRange = exports.create = void 0;
var p = require("./pos");
function create(sl, sc, el, ec) {
    return {
        start: p.create(sl, sc),
        end: p.create(el, ec)
    };
}
exports.create = create;
function fromProtocolRange(range) {
    return {
        start: p.fromProtocol(range.start),
        end: p.fromProtocol(range.end)
    };
}
exports.fromProtocolRange = fromProtocolRange;
function toProtocol(range) {
    return {
        start: p.toProtocol(range.start),
        end: p.toProtocol(range.end)
    };
}
exports.toProtocol = toProtocol;
// | ---- outer ----|
//   | -- inner --|
function isInside(inner, outer) {
    return p.isAfterOrSame(inner.start, outer.start) && p.isBeforeOrSame(inner.end, outer.end);
}
exports.isInside = isInside;
function isInsideStrict(inner, outer) {
    return p.isAfter(inner.start, outer.start) && p.isBefore(inner.end, outer.end);
}
exports.isInsideStrict = isInsideStrict;
function format(r) {
    return r.start.line + ":" + r.start.column + "-" + r.end.line + ":" + r.end.column;
}
exports.format = format;
