// yaml-language-server/out/server/test/utils/testsTypes.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestWorkspace = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
class TestWorkspace {
    applyEdit(paramOrEdit) {
        throw new Error('Method not implemented.');
    }
    getConfiguration(items) {
        throw new Error('Method not implemented.');
    }
    getWorkspaceFolders() {
        throw new Error('Method not implemented.');
    }
    onDidCreateFiles(handler) {
        throw new Error('Method not implemented.');
    }
    onDidRenameFiles(handler) {
        throw new Error('Method not implemented.');
    }
    onDidDeleteFiles(handler) {
        throw new Error('Method not implemented.');
    }
    onWillCreateFiles(handler) {
        throw new Error('Method not implemented.');
    }
    onWillRenameFiles(handler) {
        throw new Error('Method not implemented.');
    }
    onWillDeleteFiles(handler) {
        throw new Error('Method not implemented.');
    }
}
exports.TestWorkspace = TestWorkspace;
